docs/decisions/0025-stripe-event-consumption-delivery.rst

```diff
@@ -0,0 +1,133 @@
+
+Architecture of the edX backend WebHook endpoint
+------------------------------------------------
+
+#. **Ingest:** Implement a new webhook endpoint in the edX Enterprise backend.
+
+   * IDA and URL: TBD.
+
+   * The view logic should robustly ingest Stripe events, storing them in the DB.
+
+     * Use `signature validation <https://docs.stripe.com/webhooks#verify-official-libraries>`_.
+
+     * Use `de-duplication <https://docs.stripe.com/webhooks#handle-duplicate-events>`_.
+
+   * After successful ingestion, respond with HTTP 200 to tell Stripe to stop retrying sending the event.
+
+#. **Pick:** Implement an asynchronous event picker to continuously try to pick events to handle/forward.
+
+   * Architecturally, this is an infinite loop via management command running in a dedicated container/pod.
+
+   * Ingested events which have not been handled will cause ``handle_stripe_event`` tasks to be queued.
+
+   * In order to give celery time to handle the event without sacrificing loop frequency, throttle queue requests for a
+     given event ID by checking the TaskResult table just like `how we throttle tasks in enterprise-catalog
+     <https://github.com/openedx/enterprise-catalog/blob/01f5367309ee25093e414b0fd3498a48ec575073/enterprise_catalog/apps/api/tasks.py#L134>`_.
+     A recent TaskResult record for a given event is like a lock that prevents the same event from re-queuing.
+
+   * Scheduling order: In case multiple events for the same customer are pending, we should make a best effort to chain
+     the task requests for a given customer (Celery tasks can be natively chained). Handling events out of order for a
+     given customer could expose weird corner cases.
+
+#. **Handle:** Implement a celery task to "handle" an event.
+
+   * The handler task should mark the event has handled by setting the ``handled`` boolean on the event record to ``true``.
+
+   * Behavior will differ based on the event type received.
+
+Architecture of the SalesForce WebHook endpoint
+-----------------------------------------------
+
+This endpoint will be provided by the `Stripe Connector App
+<https://docs.stripe.com/plugins/stripe-connector-for-salesforce/installation-guide>`_ to directly ingest Stripe events
+within SalesForce.
+
+SalesForce Engineers will then need to trigger custom APEX code (or other automation) to handle `invoice.paid` and other
+events. As a SOQL query, recent `invoice.paid` events can be fetched like this::
+
+  select CreatedDate,
+         stripeGC__Request_Body__c
+  from stripeGC__Stripe_Event__c
+  where stripeGC__Event_Name__c = 'invoice.paid'
+    and stripeGC__Is_Live_Mode__c = true
+  order by CreatedDate desc
+  limit 10
+
+Alternatives Considered
+=======================
+
+Rejected
+========
+
+* Queue event handler immediately as part of the ingestion, instead of via an asynchronous picker.
+
+  * This is too sensitive to errors during ingestion, which could result in dropped events.
+
```

Comment:
Why not queue an _async_ task immediately as part of ingestion? What kind of errors is this sensitive to that could drop events?  Couldn't we do something like this to avoid those errors?
```python
try:
    with transaction.atomic():
        persist_the_event_data()
        transaction.on_commit(handle_the_event.apply_async)
    return Response(200)
except:
    raise # return error response
```

Reply:
If we enqueued an async task on ingestion, we could additionally run a k8s cronjob to perform the pick logic you defined above. This would serve to pick up any ingested-but-unprocessed events that may otherwise be orphaned due to celery malfunctions, etc.

Reply:
Hm, I think you're right! I probably mis-remembered the real motivation which was just that it would be less code to write/test. As long as we attempt to persist the event & COMMIT before attempting to queue it, then we can't "drop" the event even if queueing the handler crashes.